core/src/main/java/org/elasticsearch/action/support/TransportAction.java

```diff
@@ -133,6 +133,10 @@ public final void execute(Task task, Request request, ActionListener<Response> l
             return;
         }
 
+        if (task != null && request.getShouldPersistResult()) {
+            listener = new PersistentActionListener<>(taskManager, task, listener);
```

Comment:
I wonder if we can assert here that the response will always implement `ToXContent`. I think `assert action.newResponse() instanceof ToXContent` will work?